### PR TITLE
feat(2024-calcite-react): add demo app

### PR DESCRIPTION
**Pull request by rslibed**

### CODE REVIEW

File: 2024/calcite-react/demo/.eslintrc.cjs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module.exports = {
+  root: true,
+  env: { browser: true, es2020: true },
+  extends: [
+    'eslint:recommended',
+    'plugin:@typescript-eslint/recommended',
+    'plugin:react-hooks/recommended',
+  ],
+  ignorePatterns: ['dist', '.eslintrc.cjs'],
```
Comment (maxpatiiuk) [inline 2024/calcite-react/demo/.eslintrc.cjs]: (optional) `'.eslintrc.cjs'` might be ignored by eslint automatically, so this is redundant

File: 2024/calcite-react/demo/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+- [Vite](https://vitejs.dev/)
+
+## Steps to run web app locally, in terminal, run:
+
+1. `git clone https://github.com/maxpatiiuk/esri-dev-summit-presentations.git`
+2. `cd esri-dev-summit-presentations/2024/calcite-react/demo`
+3. `yarn install`
+4. `yarn start`
+
+Note: yarn and git is required run the above commands.
```
Comment (maxpatiiuk) [inline 2024/calcite-react/demo/README.md]:
```suggestion
Note: yarn and git is required to run the above commands.
```

File: 2024/calcite-react/demo/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+- [ArcGIS Maps SDK for JavaScript's ES modules](https://developers.arcgis.com/javascript/latest/)
+- [React](https://react.dev/)
+- [TypeScript](https://www.typescriptlang.org/)
+- [Vite](https://vitejs.dev/)
+
+## Steps to run web app locally, in terminal, run:
+
+1. `git clone https://github.com/maxpatiiuk/esri-dev-summit-presentations.git`
+2. `cd esri-dev-summit-presentations/2024/calcite-react/demo`
+3. `yarn install`
```
Comment (maxpatiiuk) [inline 2024/calcite-react/demo/README.md]: I'm not against yarn, but more people have `npm` installed by default
Should we use `npm install` and `npm start` here instead? They should work equivalently in this case

File: 2024/calcite-react/demo/src/Components/View/View.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import MapView from "@arcgis/core/views/MapView";
+
+import "./View.scss";
+
+interface ViewProps {
+    webmap: string;
+}
+
+export const View = (props: ViewProps) => {
+    const viewRef = useRef(null) as React.RefObject<HTMLDivElement>;
```
Comment (maxpatiiuk) [inline 2024/calcite-react/demo/src/Components/View/View.tsx]: to reduce type assertions:

```suggestion
    const viewRef = useRef<HTMLDivElement | null>(null);
```

File: 2024/calcite-react/demo/src/App.scss [hunk 5]
```diff
@@ -0,0 +1,8 @@
+header {
```
Comment (maxpatiiuk) [inline 2024/calcite-react/demo/src/App.scss]: oh, I didn't realize that vite has scss support out of the box - that's cool

(optional) though we could still use regular .css here to make the demo app more approachable for people not familiar with scss

File: 2024/calcite-react/demo/src/main.tsx [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
+import ReactDOM from 'react-dom/client';
+import App from './App.tsx';
+
+import "./index.scss";
+import './App.scss';
+
+import { setAssetPath } from "@esri/calcite-components/dist/components";
+setAssetPath("https://cdn.jsdelivr.net/npm/@esri/calcite-components/dist/calcite/assets");
```
Comment (maxpatiiuk) [inline 2024/calcite-react/demo/src/main.tsx:9]: (optional) should users alway use the CDN for assets or are there use cases for using local assets?
do we need to add a comment about that here/in the presentation or is it an uncommon use case?

**Review by maxpatiiuk — APPROVED** [review]
great job!
thanks for taking care of this

File: 2024/calcite-react/README.md [hunk 7]
```diff
@@ -1,23 +1,22 @@
...
+
+When: Thursday, Mar 14 | 10:30 AM - 11:00 AM PDT
+
+Where: Oasis 1-2 | PSCC
 
 Presenters: [Ryan Libed](https://github.com/rslibed) &
 [Max Patiiuk](https://github.com/maxpatiiuk)
 
-Length: 60 minutes
+To be presented at [Esri Developer Summit 2024](http://www.esri.com/events/devsummit).
```
Comment (maxpatiiuk) [inline 2024/calcite-react/README.md]: we could link directly to the session details page on Esri's website:

```suggestion
To be presented at [Esri Developer Summit 2024](https://registration.esri.com/flow/esri/24epcdev/deveventportal/page/detailed-agenda/session/1699142656142001V31a).
```

File: 2024/calcite-react/demo/src/Components/View/View.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        });
+
+        const view = new MapView({
+            container: viewRef?.current as HTMLDivElement,
+            map
+        });
+
+        setView(view);
+
+    }, [viewRef]);
```
Comment (maxpatiiuk) [inline 2024/calcite-react/demo/src/Components/View/View.tsx]: a ref can't be used in the dependency array in React, as ref is always pointing to the same object `{ current: ... }`

if you need to use it in the dependency array, you should use a react state instead (`const [viewContainer,setViewContainer] = React.useState<HTMLDivElement | null>(null);` and `<div ref={setViewContainer} id="viewDiv" />`)

but that is only necessary if the ref element is rendered dynamically - which it isn't in this case

thus the viewRef is going to be set by the time the `useEffect` is called, and so you don't need to provide anything in the dependency array:

```suggestion
    }, []);
```

File: 2024/calcite-react/demo/src/Components/View/View.tsx [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            }
+        });
+
+        const view = new MapView({
+            container: viewRef?.current as HTMLDivElement,
+            map
+        });
+
+        setView(view);
+
```
Comment (maxpatiiuk) [inline 2024/calcite-react/demo/src/Components/View/View.tsx:42]: could you remove the useEffect above this one and inline the destruction function inside of this useEffect?

that might improve clarity a bit

```suggestion
    return () => view?.destroy();
```

File: 2024/calcite-react/demo/src/Components/View/View.tsx [hunk 10]
```diff
@@ -34,7 +34,7 @@ export const View = (props: ViewProps) => {
         });
 
         setView(view);
-
+    return () => view?.destroy();
```
Comment (maxpatiiuk) [inline 2024/calcite-react/demo/src/Components/View/View.tsx]: now that this is here, the lines 16-20 can be removed:

```ts
    useEffect(() => {
        return () => {
            view?.destroy();
        }
    }, [])
```

File: 2024/calcite-react/demo/.eslintrc.cjs (hunk 1 above)
Comment (rslibed) [reply 2024/calcite-react/demo/.eslintrc.cjs]: Done

File: 2024/calcite-react/demo/README.md (hunk 3 above)
Comment (rslibed) [reply 2024/calcite-react/demo/README.md]: Good point, updated yarn to npm

File: 2024/calcite-react/demo/src/App.scss (hunk 5 above)
Comment (rslibed) [reply 2024/calcite-react/demo/src/App.scss]: i installed sass but i'm fine either way. removed sass and converted style files to .css

File: 2024/calcite-react/demo/src/Components/View/View.tsx (hunk 10 above)
Comment (rslibed) [reply 2024/calcite-react/demo/src/Components/View/View.tsx]: Done

File: 2024/calcite-react/demo/src/main.tsx (hunk 6 above)
Comment (rslibed) [reply 2024/calcite-react/demo/src/main.tsx:9]: Based on discussion, we'll stick with what the maps sdk for js documentation recommends (using CDN). In presentation, we'll mention managing assets locally as an alternative.

**Pull request merged**